File: rust/src/bin/p0052.rs
// Project Euler: Problem 52

euler::run_solver!(52);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> u64 {
    for exp in 6_u32.. {
        for n in 10_u64.pow(exp - 1)..=(10_u64.pow(exp) / 6) {
            if check_num(n) {
                return n;
            }
        }
    }

    unreachable!();
}

fn check_num(num: u64) -> bool {
    use euler::math;

    fn make_key(n: u64) -> Vec<u64> {
        let mut key = math::digits(n);
        key.sort_unstable();
        key
    }

    let key = make_key(num);
    for mult in 2..=6 {
        if key != make_key(num * mult) {
            return false;
        }
    }

    true
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0052() {
        assert_eq!(compute(), 142857);
    }
}

File: ocaml/src/p0052.ml
(* Project Euler: Problem 52 *)

open Core

let check_num n =
  let make_key x = Euler.Util.digits x |> List.sort ~compare:Int.compare in
  let key = make_key n in
  List.(
    range 2 6 ~stop:`inclusive
    |> for_all ~f:(fun i -> equal Int.equal key (make_key (n * i))))
;;

let compute () =
  Sequence.(
    unfold ~init:6 ~f:(fun n -> Some (n, n + 1))
    |> find_map ~f:(fun exp ->
      range (Int.pow 10 (exp - 1)) (Int.pow 10 exp / 6) ~stop:`inclusive
      |> find ~f:check_num)
    |> Option.value_exn)
;;

let solve () = compute () |> Int.to_string

(* Test *)

let%test_unit "Permuted Multiples" = [%test_eq: int] (compute ()) 142857
